### PR TITLE
feat(sw): use service worker in development mode

### PR DESCRIPTION
**Pull request by KaiVandivier**
Enables apps to use a service worker while using `yarn start` to run a development server. It compiles a service worker and adds it to the `public` dir of the app shell where it can be registered.

In development mode, nothing will be served cache-first so that updated files are used. Everything will still be cached network-first (unless filtered) so that it can still be used offline though.

This compilation script also opens up the door for using it in for the production builds too, which gives us more control over the `injectManifest` configuration, the bundling options, and environment variable usage

### CODE REVIEW

File: cli/src/lib/pwa/getPWAEnvVars.js [hunk 1]
```diff
@@ -7,8 +7,10 @@
 function getPWAEnvVars(config) {
     if (config.type !== 'app' || !config.pwa.enabled) return null
     return {
-        pwa_enabled: config.pwa.enabled,
-        omit_external_requests: config.pwa.caching.omitExternalRequests,
+        pwa_enabled: JSON.stringify(config.pwa.enabled),
+        omit_external_requests: JSON.stringify(
+            config.pwa.caching.omitExternalRequests
+        ),
```
Comment (KaiVandivier) [inline cli/src/lib/pwa/getPWAEnvVars.js:13]: These are now all stringified for adding to the dev SW env

File: cli/src/lib/shell/index.js [hunk 2]
```diff
@@ -28,7 +25,7 @@ module.exports = ({ config, paths }) => ({
             cmd: 'yarn',
             args: ['run', 'start'],
             cwd: paths.shell,
-            env: getEnv({ name: config.title, port }),
+            env: getEnv({ name: config.title, port, ...getPWAEnvVars(config) }),
```
Comment (KaiVandivier) [inline cli/src/lib/shell/index.js:28]: Adds PWA env vars to app env when using `start` script

File: sw/src/lib/registration.js [hunk 3]
```diff
@@ -49,7 +49,7 @@ export function register(config) {
             )
     )
 
-    if (process.env.NODE_ENV === 'production' && 'serviceWorker' in navigator) {
+    if ('serviceWorker' in navigator) {
```
Comment (KaiVandivier) [inline sw/src/lib/registration.js:52]: Enables SW registration outside of production environments

File: sw/src/lib/sections-db.js [hunk 4]
```diff
@@ -25,7 +25,7 @@ export function openSectionsDB() {
                 }
                 // falls through (this comment satisfies eslint)
                 default: {
-                    console.log('[sections-db] Done upgrading DB')
+                    console.debug('[sections-db] Done upgrading DB')
```
Comment (KaiVandivier) [inline sw/src/lib/sections-db.js:28]: (shh)

File: sw/src/service-worker/service-worker.js [hunk 5]
```diff
@@ -21,6 +21,7 @@ function setUpKillSwitchServiceWorker() {
     })
 
     self.addEventListener('activate', async () => {
+        console.log('Removing previous service worker')
```
Comment (KaiVandivier) [inline sw/src/service-worker/service-worker.js:24]: It's pretty important to know that the killswitch worker is getting used

File: sw/src/service-worker/service-worker.js [hunk 6]
```diff
@@ -124,9 +134,11 @@ export function setUpServiceWorker() {
...
     // If not recording, fall through to default caching strategies
-    // SWR strategy for static assets that can't be precached
+    // SWR strategy for static assets that can't be precached.
+    // Skip in development environments
     registerRoute(
         ({ url }) =>
+            PRODUCTION_ENV &&
             urlMeetsDefaultCachingCriteria(url) &&
             fileExtensionRegexp.test(url.pathname),
         new StaleWhileRevalidate({ cacheName: 'other-assets' })
```
Comment (KaiVandivier) [inline sw/src/service-worker/service-worker.js:144]: So newest static assets are available ASAP in development mode, skip SWR strategy for static assets

Comment (KaiVandivier) [inline sw/src/service-worker/service-worker.js:144]: So newest static assets are available ASAP in development mode, skip SWR strategy for static asset

**Review by mediremi — APPROVED** [review]
LGTM :heavy_check_mark:

**Pull request merged**